packages/jellyfish-testing/src/testing.ts

```diff
@@ -6,24 +6,31 @@ import { TestingRawTx } from './rawtx'
 import { TestingICX } from './icxorderbook'
 import { TestingMisc } from './misc'
 import { TestingGroupAnchor } from './anchor'
-import { ContainerGroup, MasterNodeRegTestContainer, StartOptions } from '@defichain/testcontainers'
+import { ContainerGroup, DeFiDContainer, MasterNodeRegTestContainer, StartOptions } from '@defichain/testcontainers'
 import { RegTestFoundationKeys } from '@defichain/jellyfish-network'
 import { JsonRpcClient } from '@defichain/jellyfish-api-jsonrpc'
 
 export class Testing {
+  public readonly container: MasterNodeRegTestContainer
   public readonly fixture = new TestingFixture(this)
-  public readonly token = new TestingToken(this.container, this.rpc)
-  public readonly poolpair = new TestingPoolPair(this.container, this.rpc)
-  public readonly rawtx = new TestingRawTx(this.container, this.rpc)
-  public readonly icxorderbook = new TestingICX(this)
-  public readonly misc = new TestingMisc(this.container, this.rpc)
+  public readonly token: TestingToken
+  public readonly poolpair: TestingPoolPair
+  public readonly rawtx: TestingRawTx
+  public readonly icxorderbook: TestingICX
+  public readonly misc: TestingMisc
 
   private readonly addresses: Record<string, string> = {}
 
   private constructor (
-    public readonly container: MasterNodeRegTestContainer,
+    public readonly con: DeFiDContainer,
     public readonly rpc: TestingJsonRpcClient
   ) {
+    this.container = con as MasterNodeRegTestContainer
```

Comment:
coming from a cpp background don't like this downcasting. especially given that we might pass instances that are children of `DeFiDContainer`, but not `MasterNodeRegTestContainer`. i.e `RegTestContainer`
In such a case extra properties of the ``MasterNodeRegTestContainer`` is not available for the casted object.
 Maybe we can check the instance type and not define `token, poolpair, ...` that depends on `MasterNodeRegTestContainer` extra properties.

Reply:
this is indeed a 'dirty' method. plus we have the `generate()` which will only work for `MasterNodeRegTestContainer`

Reply:
i have resulted in creating `generate` and `generatenewaddress` method in DefiDcontainer. also checking for the instance of `this.container` before initiating` TestToken,...` and the like. it requires me to forcibly assert the variables to be non null which is something that i dont quite like however

Reply:
Here is one way we can handle this, its dirty but it gives a decent amount of intellisense/inference of the typing.

```ts
export class Testing<Container extends DeFiDContainer = DeFiDContainer> {
  public readonly fixture: TestingFixture
  public readonly icxorderbook: TestingICX

  public readonly token!: Container extends MasterNodeRegTestContainer ? TestingToken : undefined
  public readonly poolpair!: Container extends MasterNodeRegTestContainer ? TestingPoolPair : undefined
  public readonly rawtx!: Container extends MasterNodeRegTestContainer ? TestingRawTx : undefined
  public readonly misc!: Container extends MasterNodeRegTestContainer ? TestingMisc : undefined

  private readonly addresses: Record<string, string> = {}

  private constructor (
    public readonly container: Container,
    public readonly rpc: TestingJsonRpcClient
  ) {
    this.fixture = new TestingFixture(this)
    this.icxorderbook = new TestingICX(this)
    if (container instanceof MasterNodeRegTestContainer) {
      this.token = new TestingToken(container, this.rpc) as any
      this.poolpair = new TestingPoolPair(container, this.rpc) as any
      this.rawtx = new TestingRawTx(container, this.rpc) as any
      this.misc = new TestingMisc(container, this.rpc) as any
    }
  }
  
  ...
  
  static create<Container extends DeFiDContainer> (container: Container): Testing<Container> {
    return this.createBase(container)
  }

  static createBase<Container extends DeFiDContainer> (container: Container = new DeFiDContainer('regtest') as Container): Testing<Container> {
    const rpc = new TestingJsonRpcClient(container)
    return new Testing(container, rpc)
  }
```

With this the Testing context is based on the container provided during creation.

![inference-2](https://user-images.githubusercontent.com/1998130/152943630-236c5d12-0275-4c54-849b-cec0035af6c9.png)

When passing a non master node container we get typescript feedback that certain functionality is not available.

![inference-1](https://user-images.githubusercontent.com/1998130/152943619-c3a1408d-1910-40ba-ba97-a6bbbd6e0a57.png)

When passing master node all is well.

This should help inform development efforts when we are writing tests. There is a couple of dirty work arounds required, like casting to `any` on the instanceof. And still needing to assert but now we can at least say they assert to undefined if not a MasterNode.

What do you think?

Reply:
i think it is an improvement since it allow intellisense to help

Reply:
`public readonly container: Container` results in some complains that `Argument of type 'DeFiDContainer' is not assignable to parameter of type 'MasterNodeRegTestContainer'.` unless all instances are changed, it seems that this method may not work? or am i missing something?

Reply:
Could you point me to where this complaint is occurring as I do not see it on my end at the moment?

Reply:
my bad i probably did a mistake as im no longer seeing that specific complain. however there is another one that is occuring in the CI now

Reply:
reverting to my orignal solution for now as after tinkering with this method. it seems that there are a couple of things that may have conflicts. 
1. the fixture file is complaining that there is a potential undefined object and trying to workaround it may not be elegant
2. seems like we are not able to rely on intellisense when using the testing wrapper to create the Testing instance

Reply:
I think for now getting the wrapper in takes priority and we can refactor the typing in a second round of refactoring perhaps?

Reply:
yea that seems like the low hanging fruit for now without affecting too many files